sdk/python/kfp/compiler/pipeline_spec_builder.py

```diff
@@ -948,22 +936,19 @@ def build_task_spec_for_group(
                     input_name].component_input_artifact = (
                         channel_full_name
                         if is_parent_component_root else input_name)
+        elif channel.task_name and channel.task_name in tasks_in_current_dag:
+            pipeline_task_spec.inputs.parameters[
+                input_name].task_output_parameter.producer_task = (
+                    utils.sanitize_task_name(channel.task_name))
+            pipeline_task_spec.inputs.parameters[
+                input_name].task_output_parameter.output_parameter_key = (
+                    channel_name)
         else:
-            # channel is one of PipelineParameterChannel, LoopArgument, or
```

Comment:
The original code would be easier to follow as the outmost if-else is artifacts vs. parameters (with loop arguments being special parameters), then under each case there're sub-conditions as upstream within the dag vs. outside.
That being said, give loop will start to accept artifacts soon, the logic here will likely change.